File: 024__Swap_Nodes_in_Pairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return head

        dummy = ListNode(0)
        dummy.next = head
        prev, curr = dummy, head

        while curr:
            first, second = curr, curr.next
            if not second:
                break
            third = second.next
            prev.next = second
            second.next = first
            first.next = third

            prev, curr = curr, curr.next
        return dummy.next

File: Java/024__Swap_Nodes_in_Pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode result = null;
        if (head == null) {
            return result;
        } else if (head.next == null) {
            return head;
        }
        ListNode first = head, second = head.next, prev = new ListNode(-1);

        while (first != null && second != null) {
            first.next = second.next;
            second.next = first;
            prev.next = second;
            if (result == null) {
                result = second;
            }
            prev = first;
            first = first.next;
            if (first == null) {
                break;
            }
            second = first.next;
        }
        return result;
    }
}
